File: server/ai/reinforcement.py
import math
import random
from typing import Dict, List, Any, Optional, Tuple
from server.game.ballistics import PhysicsConfig, simulate_trajectory

class QLearningAgent:
    def __init__(
        self,
        angle_bins: int = 18,
        power_bins: int = 10,
        learning_rate: float = 0.1,
        discount_factor: float = 0.95,
        epsilon: float = 0.2
    ):
        self.angle_bins = angle_bins
        self.power_bins = power_bins
        self.learning_rate = learning_rate
        self.discount_factor = discount_factor
        self.epsilon = epsilon
        self.q_table: Dict[str, float] = {}
        self.experience_buffer: List[Dict] = []
    
    def _discretize_state(self, origin: Dict[str, float], target_pos: Dict[str, float], wind: Dict[str, float]) -> str:
        dx = round((target_pos['x'] - origin['x']) / 5) * 5
        dy = round((target_pos['y'] - origin['y']) / 2) * 2
        wind_bucket = round(wind['speed'] / 3) * 3
        return f"{dx}_{dy}_{wind_bucket}"
    
    def _action_to_params(self, action: int) -> Tuple[float, float]:
        angle_idx = action // self.power_bins
        power_idx = action % self.power_bins
        
        angle = 10 + (angle_idx * 160 / (self.angle_bins - 1))
        power = 20 + (power_idx * 80 / (self.power_bins - 1))
        
        return angle, power
    
    def _get_q_value(self, state: str, action: int) -> float:
        key = f"{state}_{action}"
        return self.q_table.get(key, 0.0)
    
    def _set_q_value(self, state: str, action: int, value: float):
        key = f"{state}_{action}"
        self.q_table[key] = value
    
    def get_action(
        self,
        origin: Dict[str, float],
        target_pos: Dict[str, float],
        wind: Dict[str, float]
    ) -> Dict[str, Any]:
        state = self._discretize_state(origin, target_pos, wind)
        
        if random.random() < self.epsilon:
            action = random.randint(0, self.angle_bins * self.power_bins - 1)
            exploration = True
        else:
            best_action = 0
            best_value = float('-inf')
            for a in range(self.angle_bins * self.power_bins):
                value = self._get_q_value(state, a)
                if value > best_value:
                    best_value = value
                    best_action = a
            action = best_action
            exploration = False
        
        angle, power = self._action_to_params(action)
        
        return {
            'angle': angle,
            'power': power,
            'action': action,
            'state': state,
            'exploration': exploration,
            'q_value': self._get_q_value(state, action)
        }
    
    def update(self, state: str, action: int, reward: float, next_state: str, done: bool):
        current_q = self._get_q_value(state, action)
        
        if done:
            target_q = reward
        else:
            max_next_q = max(self._get_q_value(next_state, a) 
                           for a in range(self.angle_bins * self.power_bins))
            target_q = reward + self.discount_factor * max_next_q
        
        new_q = current_q + self.learning_rate * (target_q - current_q)
        self._set_q_value(state, action, new_q)
    
    def get_stats(self) -> Dict[str, Any]:
        return {
            'q_table_size': len(self.q_table),
            'epsilon': self.epsilon,
            'experience_count': len(self.experience_buffer)
        }


class SupervisedAgent:
    def __init__(self, angle_resolution: int = 5, power_resolution: int = 5):
        self.angle_resolution = angle_resolution
        self.power_resolution = power_resolution
    
    def get_optimal_shot(
        self,
        origin: Dict[str, float],
        target_box: Dict[str, float],
        max_velocity: float,
        physics: PhysicsConfig,
        terrain_height: float
    ) -> Dict[str, Any]:
        best_angle = 45
        best_power = 50
        best_score = float('-inf')
        
        target_center = {
            'x': (target_box['min_x'] + target_box['max_x']) / 2,
            'y': (target_box['min_y'] + target_box['max_y']) / 2
        }
        
        for angle in range(10, 171, self.angle_resolution):
            for power in range(20, 101, self.power_resolution):
                result = simulate_trajectory(
                    origin=origin,
                    angle=angle,
                    power=power,
                    max_velocity=max_velocity,
                    physics=physics,
                    terrain_height=terrain_height,
                    target_box=target_box
                )
                
                if result['hit_target']:
                    score = 1000 + result['damage']
                else:
                    final_pos = result['final_position']
                    distance_to_target = math.sqrt(
                        (final_pos['x'] - target_center['x'])**2 +
                        (final_pos['y'] - target_center['y'])**2
                    )
                    score = -distance_to_target
                
                if score > best_score:
                    best_score = score
                    best_angle = angle
                    best_power = power
        
        return {
            'angle': best_angle,
            'power': best_power,
            'score': best_score,
            'is_hit': best_score >= 1000
        }

File: client/src/lib/ai/reinforcement.ts
// Reinforcement Learning module for Autopilot Kitten
// Implements Q-learning policy for aiming decisions

import { 
  Vector2D, 
  PhysicsConfig, 
  simulateTrajectory,
  calculateDamage,
  CollisionBox,
} from "../physics/ballistics";

export interface RLState {
  playerX: number;
  playerY: number;
  opponentX: number;
  opponentY: number;
  wind: number;
  playerHealth: number;
  opponentHealth: number;
}

export interface RLAction {
  angle: number;  // 0-180 degrees
  power: number;  // 0-100 percent
}

export interface Experience {
  state: RLState;
  action: RLAction;
  reward: number;
  nextState: RLState;
  done: boolean;
}

export interface PolicyNetwork {
  weights: number[][][];
  biases: number[][];
}

export interface RLConfig {
  learningRate: number;
  discountFactor: number;
  explorationRate: number;
  explorationDecay: number;
  minExploration: number;
  batchSize: number;
}

export const DEFAULT_RL_CONFIG: RLConfig = {
  learningRate: 0.001,
  discountFactor: 0.99,
  explorationRate: 1.0,
  explorationDecay: 0.995,
  minExploration: 0.01,
  batchSize: 32,
};

// Discretize state for tabular Q-learning
function discretizeState(state: RLState): string {
  const windBucket = Math.round(state.wind / 2);
  const distX = Math.round((state.opponentX - state.playerX) / 5);
  const distY = Math.round((state.opponentY - state.playerY) / 2);
  
  return `${distX},${distY},${windBucket}`;
}

// Discretize action space
function discretizeAction(action: RLAction): string {
  const angleBucket = Math.round(action.angle / 5);
  const powerBucket = Math.round(action.power / 10);
  
  return `${angleBucket},${powerBucket}`;
}

// Parse discretized action back to continuous
function parseAction(actionKey: string): RLAction {
  const [angleBucket, powerBucket] = actionKey.split(",").map(Number);
  return {
    angle: angleBucket * 5,
    power: powerBucket * 10,
  };
}

// Q-Learning Agent for Autopilot Kitten
export class QLearningAgent {
  private qTable: Map<string, Map<string, number>>;
  private config: RLConfig;
  private explorationRate: number;
  private replayBuffer: Experience[];
  private trainingStats: { episode: number; reward: number; wins: number }[];

  constructor(config: RLConfig = DEFAULT_RL_CONFIG) {
    this.qTable = new Map();
    this.config = config;
    this.explorationRate = config.explorationRate;
    this.replayBuffer = [];
    this.trainingStats = [];
  }

  // Get Q-value for state-action pair
  private getQValue(stateKey: string, actionKey: string): number {
    const stateQ = this.qTable.get(stateKey);
    if (!stateQ) return 0;
    return stateQ.get(actionKey) || 0;
  }

  // Set Q-value for state-action pair
  private setQValue(stateKey: string, actionKey: string, value: number): void {
    if (!this.qTable.has(stateKey)) {
      this.qTable.set(stateKey, new Map());
    }
    this.qTable.get(stateKey)!.set(actionKey, value);
  }

  // Get all possible actions (discretized)
  private getActionSpace(): string[] {
    const actions: string[] = [];
    for (let angle = 0; angle <= 180; angle += 5) {
      for (let power = 10; power <= 100; power += 10) {
        actions.push(`${angle / 5},${power / 10}`);
      }
    }
    return actions;
  }

  // Select action using epsilon-greedy policy
  selectAction(state: RLState): RLAction {
    const stateKey = discretizeState(state);
    
    // Exploration
    if (Math.random() < this.explorationRate) {
      return {
        angle: Math.random() * 180,
        power: 10 + Math.random() * 90,
      };
    }
    
    // Exploitation - find best action
    const actions = this.getActionSpace();
    let bestAction = actions[0];
    let bestValue = -Infinity;
    
    for (const action of actions) {
      const value = this.getQValue(stateKey, action);
      if (value > bestValue) {
        bestValue = value;
        bestAction = action;
      }
    }
    
    return parseAction(bestAction);
  }

  // Update Q-value using TD learning
  update(experience: Experience): void {
    this.replayBuffer.push(experience);
    
    const stateKey = discretizeState(experience.state);
    const actionKey = discretizeAction(experience.action);
    
    // Calculate target
    let target = experience.reward;
    if (!experience.done) {
      const nextStateKey = discretizeState(experience.nextState);
      const actions = this.getActionSpace();
      let maxNextQ = -Infinity;
      
      for (const action of actions) {
        const value = this.getQValue(nextStateKey, action);
        if (value > maxNextQ) {
          maxNextQ = value;
        }
      }
      
      target += this.config.discountFactor * maxNextQ;
    }
    
    // Update Q-value
    const currentQ = this.getQValue(stateKey, actionKey);
    const newQ = currentQ + this.config.learningRate * (target - currentQ);
    this.setQValue(stateKey, actionKey, newQ);
    
    // Decay exploration
    this.explorationRate = Math.max(
      this.config.minExploration,
      this.explorationRate * this.config.explorationDecay
    );
  }

  // Calculate reward from shot result
  static calculateReward(
    hit: boolean,
    damage: number,
    opponentHealthBefore: number,
    opponentHealthAfter: number,
    playerWon: boolean
  ): number {
    let reward = 0;
    
    if (hit) {
      reward += 10;
      reward += damage * 0.5;
    } else {
      reward -= 1;
    }
    
    if (playerWon) {
      reward += 50;
    }
    
    if (opponentHealthAfter <= 0) {
      reward += 30;
    }
    
    return reward;
  }

  // Get training statistics
  getStats(): typeof this.trainingStats {
    return this.trainingStats;
  }

  // Record episode stats
  recordEpisode(reward: number, won: boolean): void {
    this.trainingStats.push({
      episode: this.trainingStats.length + 1,
      reward,
      wins: won ? 1 : 0,
    });
  }

  // Get current exploration rate
  getExplorationRate(): number {
    return this.explorationRate;
  }

  // Export policy for persistence
  exportPolicy(): object {
    const policy: Record<string, Record<string, number>> = {};
    this.qTable.forEach((actions, state) => {
      policy[state] = {};
      actions.forEach((value, action) => {
        policy[state][action] = value;
      });
    });
    return policy;
  }

  // Import policy
  importPolicy(policy: Record<string, Record<string, number>>): void {
    this.qTable.clear();
    for (const state in policy) {
      this.qTable.set(state, new Map(Object.entries(policy[state])));
    }
  }
}

// Supervised baseline agent - learns from optimal solver
export class SupervisedAgent {
  private solutionCache: Map<string, RLAction>;

  constructor() {
    this.solutionCache = new Map();
  }

  // Calculate optimal shot using physics simulation
  findOptimalShot(
    playerPos: Vector2D,
    targetBox: CollisionBox,
    physicsConfig: PhysicsConfig,
    terrainHeight: number,
    maxVelocity: number
  ): RLAction {
    const cacheKey = `${playerPos.x},${playerPos.y},${targetBox.x},${targetBox.y},${physicsConfig.windSpeed}`;
    
    if (this.solutionCache.has(cacheKey)) {
      return this.solutionCache.get(cacheKey)!;
    }
    
    let bestAction: RLAction = { angle: 45, power: 50 };
    let bestScore = -Infinity;
    
    // Grid search for optimal angle and power
    for (let angle = 10; angle <= 170; angle += 2) {
      for (let power = 20; power <= 100; power += 5) {
        const result = simulateTrajectory(
          { origin: playerPos, angle, power, maxVelocity },
          physicsConfig,
          terrainHeight,
          targetBox
        );
        
        let score = 0;
        if (result.hitTarget) {
          score = 100 + calculateDamage(result.impactVelocity);
        } else {
          // Score based on how close we got
          const targetCenter = {
            x: targetBox.x + targetBox.width / 2,
            y: targetBox.y + targetBox.height / 2,
          };
          const dist = Math.sqrt(
            (result.finalPosition.x - targetCenter.x) ** 2 +
            (result.finalPosition.y - targetCenter.y) ** 2
          );
          score = Math.max(0, 50 - dist);
        }
        
        if (score > bestScore) {
          bestScore = score;
          bestAction = { angle, power };
        }
      }
    }
    
    this.solutionCache.set(cacheKey, bestAction);
    return bestAction;
  }

  // Select action (deterministic - always picks optimal)
  selectAction(
    state: RLState,
    physicsConfig: PhysicsConfig,
    terrainHeight: number,
    maxVelocity: number
  ): RLAction {
    const playerPos = { x: state.playerX, y: state.playerY };
    const targetBox: CollisionBox = {
      x: state.opponentX - 0.5,
      y: state.opponentY - 0.5,
      width: 1,
      height: 2,
    };
    
    return this.findOptimalShot(
      playerPos,
      targetBox,
      physicsConfig,
      terrainHeight,
      maxVelocity
    );
  }
}

// Policy comparison utilities
export interface PolicyComparison {
  rlWins: number;
  supervisedWins: number;
  rlAvgReward: number;
  supervisedAvgReward: number;
  totalGames: number;
}

export function comparePolicies(
  rlStats: { reward: number; wins: number }[],
  supervisedStats: { reward: number; wins: number }[]
): PolicyComparison {
  const rlTotal = rlStats.length;
  const supervisedTotal = supervisedStats.length;
  
  return {
    rlWins: rlStats.filter(s => s.wins > 0).length,
    supervisedWins: supervisedStats.filter(s => s.wins > 0).length,
    rlAvgReward: rlStats.reduce((sum, s) => sum + s.reward, 0) / (rlTotal || 1),
    supervisedAvgReward: supervisedStats.reduce((sum, s) => sum + s.reward, 0) / (supervisedTotal || 1),
    totalGames: Math.max(rlTotal, supervisedTotal),
  };
}
